pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -809,17 +827,18 @@ func (gc *GarbageCollector) GraphHasUID(u types.UID) bool {
 // All discovery errors are considered temporary. Upon encountering any error,
 // GetDeletableResources will log and return any discovered resources it was
 // able to process (which may be none).
-func GetDeletableResources(discoveryClient discovery.ServerResourcesInterface) map[schema.GroupVersionResource]struct{} {
+func GetDeletableResources(discoveryClient discovery.ServerResourcesInterface) (fetchedResources map[schema.GroupVersionResource]struct{}, groupLookupFailures map[schema.GroupVersion]error) {
 	preferredResources, err := discoveryClient.ServerPreferredResources()
 	if err != nil {
 		if discovery.IsGroupDiscoveryFailedError(err) {
 			klog.Warningf("failed to discover some groups: %v", err.(*discovery.ErrGroupDiscoveryFailed).Groups)
+			groupLookupFailures = err.(*discovery.ErrGroupDiscoveryFailed).Groups
 		} else {
 			klog.Warningf("failed to discover preferred resources: %v", err)
 		}
 	}
 	if preferredResources == nil {
-		return map[schema.GroupVersionResource]struct{}{}
+		return map[schema.GroupVersionResource]struct{}{}, groupLookupFailures
```

Comment:
Hmm.. preferedResources could be `nil` due to err being non-nil, in which case we don't return _any_ `fetchedResources` nor `groupLookupFailures`, nor do we indicate an error has occurred to the caller... I guess this is all ok though, due to the explicit check for a zero length `fetchedResources` which in turn skips the rest of the Sync, thus we won't remove any handlers.

So.. I think this is fine/isn't something we need to adapt for. I'm not sure how much you want to tie the implementation of this function to the call-site (`Sync`), but it may be worth adding a little note to that effect here for the next person that reads this?

Reply:
switched this to just return the error and extract the failed groups at the call site, just like the resource quota controller did